src/sycltest/plugin-Test2/gpuAlgo2.cc

```diff
@@ -0,0 +1,202 @@
+
+  auto d_c = cms::sycltools::make_device_unique<float[]>(NUM_VALUES, stream);
+
+#ifdef DEBUG
+  std::cerr << "block size: " << threadsPerBlock << " / " << maxWorkgroupSize << std::endl;
+  std::cerr << "grid size:  " << blocksPerGrid << std::endl;
+#endif
+  stream.submit([&](sycl::handler &cgh) {
+    auto pd_a = d_a.get();
+    auto pd_b = d_b.get();
+    auto pd_c = d_c.get();
+
+    cgh.parallel_for(grid, [=](sycl::nd_item<1> item) { vectorAdd(pd_a, pd_b, pd_c, NUM_VALUES, item); });
+  });
+
+  auto d_ma = cms::sycltools::make_device_unique<float[]>(NUM_VALUES * NUM_VALUES, stream);
+  auto d_mb = cms::sycltools::make_device_unique<float[]>(NUM_VALUES * NUM_VALUES, stream);
+  auto d_mc = cms::sycltools::make_device_unique<float[]>(NUM_VALUES * NUM_VALUES, stream);
+
+  sycl::range<3> threadsPerBlock3{1, workgroupRectB, workgroupRectA};
+  sycl::range<3> blocksPerGrid3{1, make_blocks(NUM_VALUES, workgroupRectB), make_blocks(NUM_VALUES, workgroupRectA)};
+  auto threadsPerGrid3 = blocksPerGrid3 * threadsPerBlock3;
+  auto grid3 = sycl::nd_range<3>(threadsPerGrid3, threadsPerBlock3);
+
+#ifdef DEBUG
+  std::cerr << "block size: " << threadsPerBlock3 << " / " << maxWorkgroupSize << std::endl;
+  std::cerr << "grid size:  " << blocksPerGrid3 << std::endl;
+#endif
+  stream.submit([&](sycl::handler &cgh) {
+    auto pd_a = d_a.get();
+    auto pd_b = d_b.get();
+    auto pd_ma = d_ma.get();
+
+    cgh.parallel_for(grid3, [=](sycl::nd_item<3> item) { vectorProd(pd_a, pd_b, pd_ma, NUM_VALUES, item); });
+  });
+
+#ifdef DEBUG
+  std::cerr << "block size: " << threadsPerBlock3 << " / " << maxWorkgroupSize << std::endl;
+  std::cerr << "grid size:  " << blocksPerGrid3 << std::endl;
+#endif
+  stream.submit([&](sycl::handler &cgh) {
+    auto pd_a = d_a.get();
+    auto pd_c = d_c.get();
+    auto pd_mb = d_mb.get();
+
+    cgh.parallel_for(grid3, [=](sycl::nd_item<3> item) { vectorProd(pd_a, pd_c, pd_mb, NUM_VALUES, item); });
+  });
+
+#ifdef DEBUG
+  std::cerr << "block size: " << threadsPerBlock3 << " / " << maxWorkgroupSize << std::endl;
+  std::cerr << "grid size:  " << blocksPerGrid3 << std::endl;
+#endif
+  stream.submit([&](sycl::handler &cgh) {
+    auto pd_ma = d_ma.get();
+    auto pd_mb = d_mb.get();
+    auto pd_mc = d_mc.get();
+
+    cgh.parallel_for(grid3, [=](sycl::nd_item<3> item) { matrixMul(pd_ma, pd_mb, pd_mc, NUM_VALUES, item); });
+  });
+
+#ifdef DEBUG
+  std::cerr << "block size: " << threadsPerBlock << " / " << maxWorkgroupSize << std::endl;
+  std::cerr << "grid size:  " << blocksPerGrid << std::endl;
+#endif
+  stream.submit([&](sycl::handler &cgh) {
+    auto pd_mc = d_mc.get();
+    auto pd_b = d_b.get();
+    auto pd_c = d_c.get();
+
+    cgh.parallel_for(grid, [=](sycl::nd_item<1> item) { matrixMulVector(pd_mc, pd_b, pd_c, NUM_VALUES, item); });
+  });
+
+  // wait for all asynchronous work to complete before destroying the device memory buffers
+  stream.submit_barrier().wait();
```

Comment:
What happens if this synchronization is not done? Program crashes?

Reply:
Yes... segmentation fault in one of the kernels.

Reply:
I opened #106 to discuss how to remove that synchronization.